python/sdist/amici/sbml_import.py

```diff
@@ -2785,58 +2801,156 @@ def _clean_reserved_symbols(self) -> None:
+        :raises SBMLException:
+            In case of unsupported expressions.
+        :returns:
+            The sympy expression or None if the input was None or an empty string.
         """
-        if isinstance(var_or_math, libsbml.SBase):
-            math_string = libsbml.formulaToL3StringWithSettings(
-                var_or_math.getMath(), self.sbml_parser_settings
+        if var_or_math is None:
+            return None
+
+        # numeric values
+        if isinstance(var_or_math, float | int):
+            return (
+                sp.Integer(var_or_math)
+                if var_or_math.is_integer()
+                else sp.Float(var_or_math)
             )
-            ele_name = var_or_math.element_name
-        else:
-            math_string = var_or_math
-            ele_name = "string"
-        math_string = replace_logx(math_string)
-        try:
+
+        # already a sympy object
+        if isinstance(var_or_math, sp.Basic):
+            # substitute free symbols to match assumptions of other model
+            #  entities where necessary
+            ele_name = "SymPy expression"
+            expr = var_or_math.subs(
+                {
+                    sym: local
+                    for sym in var_or_math.free_symbols
+                    if (local := self._local_symbols.get(str(sym), sym)) != sym
+                }
+            )
+
+        # an expression string
+        elif isinstance(var_or_math, str):
+            if not var_or_math:
+                return None
+
+            ele_name = "Expression string"
             try:
-                formula = sp.sympify(
-                    _parse_logical_operators(math_string),
-                    locals=self._local_symbols,
+                expr = sp.sympify(
+                    var_or_math,
+                    locals={
+                        "true": sp.Float(1.0),
+                        "false": sp.Float(0.0),
+                        **self._local_symbols,
+                    },
                 )
-            except TypeError as err:
-                if str(err) == "BooleanAtom not allowed in this context.":
-                    formula = sp.sympify(
-                        _parse_logical_operators(math_string),
-                        locals={
-                            "true": sp.Float(1.0),
-                            "false": sp.Float(0.0),
-                            **self._local_symbols,
-                        },
-                    )
-                else:
-                    raise
-        except (sp.SympifyError, TypeError, ZeroDivisionError) as err:
-            raise SBMLException(
-                f'{ele_name} "{math_string}" '
-                "contains an unsupported expression: "
-                f"{err}."
+            except (sp.SympifyError, TypeError, ZeroDivisionError) as err:
+                raise SBMLException(
+                    f'{ele_name} "{var_or_math}" '
```

Comment:
this will never be triggered from ele_name == "SymPy expression", right? Should the try claus also include "isinstance(var_or_math, sp.Basic)"?

Reply:
> this will never be triggered from ele_name == "SymPy expression", right? 

right. that ele_name can be removed.

> Should the try claus also include "isinstance(var_or_math, sp.Basic)"?

I don't think so. I don't recall errors during `.subs()`